File: modules/data/dmetro.lua
sptbl["dmetro"] = {

    files = {
        module = "dmetro.c",
        header = "dmetro.h",
        example = "ex_dmetro.c",
    },

    func = {
        create = "ut_dmetro_create",
        destroy = "ut_dmetro_destroy",
        init = "ut_dmetro_init",
        compute = "ut_dmetro_compute",
    },

    params = {
        optional = {
            {
                name = "time",
                type = "UTFLOAT",
                description ="Time between triggers (in seconds). This will update at the start of each trigger.",
                default = 1.0
            },
        }
    },

    modtype = "module",

    description = [[Delta Metro

    Produce a set of triggers spaced apart by time.]],

    ninputs = 0,
    noutputs = 1,

    inputs = {
        {
            name = "dummy",
            description = "this doesn't do anything."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Trigger output."
        },
    }

}

File: modules/dmetro.c
#include <stdlib.h>
#include "utone.h"

int ut_dmetro_create(ut_dmetro **p)
{
    *p = malloc(sizeof(ut_dmetro));
    return UT_OK;
}

int ut_dmetro_destroy(ut_dmetro **p)
{
    free(*p);
    return UT_OK;
}

int ut_dmetro_init(ut_data *ut, ut_dmetro *p)
{
    p->counter = 0;
    p->time = 1.0;
    return UT_OK;
}

int ut_dmetro_compute(ut_data *ut, ut_dmetro *p, UTFLOAT *in, UTFLOAT *out)
{
    *out = 0; 

    if(p->counter == 0) {
        *out = 1.0;
        p->counter = (int)(ut->sr * p->time) + 1;
    }

    p->counter--; 

    return UT_OK;
}
